File: backend/app/models/pacing.py
# =============================================================================
# Stratum AI - Pacing & Forecasting Database Models
# =============================================================================
"""
Database models for targets, pacing, and forecasting.

Models:
- Target: Monthly/quarterly targets for spend, revenue, ROAS
- DailyKPI: Materialized daily KPIs for fast pacing queries
- PacingAlert: Alert records for pacing issues
- Forecast: Stored forecasts for trend analysis
"""

import enum
from datetime import date, datetime, timezone
from typing import Optional
from uuid import uuid4

from sqlalchemy import (
    BigInteger,
    Boolean,
    Column,
    Date,
    DateTime,
    Float,
    ForeignKey,
    Index,
    Integer,
    String,
    Text,
    UniqueConstraint,
)
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import relationship

from app.db.base_class import Base, StrEnumType

# =============================================================================
# Enums
# =============================================================================


class TargetPeriod(str, enum.Enum):
    """Target period types."""

    MONTHLY = "monthly"
    QUARTERLY = "quarterly"
    YEARLY = "yearly"
    CUSTOM = "custom"


class TargetMetric(str, enum.Enum):
    """Target metric types."""

    SPEND = "spend"
    REVENUE = "revenue"
    ROAS = "roas"
    CONVERSIONS = "conversions"
    LEADS = "leads"
    PIPELINE_VALUE = "pipeline_value"
    WON_REVENUE = "won_revenue"
    CPA = "cpa"
    CPL = "cpl"


class AlertSeverity(str, enum.Enum):
    """Alert severity levels."""

    INFO = "info"
    WARNING = "warning"
    CRITICAL = "critical"


class AlertType(str, enum.Enum):
    """Alert types for pacing."""

    UNDERPACING_SPEND = "underpacing_spend"
    OVERPACING_SPEND = "overpacing_spend"
    ROAS_BELOW_TARGET = "roas_below_target"
    CONVERSIONS_BELOW_TARGET = "conversions_below_target"
    REVENUE_BELOW_TARGET = "revenue_below_target"
    PIPELINE_BELOW_TARGET = "pipeline_below_target"
    PACING_CLIFF = "pacing_cliff"  # Sudden drop
    BUDGET_EXHAUSTION = "budget_exhaustion"


class AlertStatus(str, enum.Enum):
    """Alert status."""

    ACTIVE = "active"
    ACKNOWLEDGED = "acknowledged"
    RESOLVED = "resolved"
    DISMISSED = "dismissed"


# =============================================================================
# Target Model
# =============================================================================


class Target(Base):
    """
    Monthly/quarterly targets for spend, revenue, ROAS, etc.
    Supports targets at account, campaign, or platform level.
    """

    __tablename__ = "targets"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4)

    # Target identification
    name = Column(String(255), nullable=False)
    description = Column(Text, nullable=True)

    # Period
    period_type = Column(
        StrEnumType(TargetPeriod), nullable=False, default=TargetPeriod.MONTHLY
    )
    period_start = Column(Date, nullable=False)
    period_end = Column(Date, nullable=False)

    # Scope (optional - null means org-wide)
    platform = Column(String(50), nullable=True)  # meta, google, etc.
    account_id = Column(
        String(255), nullable=True
    )  # Platform account ID (e.g., "act_123456789")
    campaign_id = Column(String(255), nullable=True)
    adset_id = Column(String(255), nullable=True)

    # Target metrics
    metric_type = Column(StrEnumType(TargetMetric), nullable=False)
    target_value = Column(Float, nullable=False)
    target_value_cents = Column(BigInteger, nullable=True)  # For monetary values

    # Bounds (optional)
    min_value = Column(Float, nullable=True)  # Alert if below
    max_value = Column(Float, nullable=True)  # Alert if above

    # Alert thresholds (percentage deviation)
    warning_threshold_pct = Column(Float, default=10.0)  # Warn at 10% deviation
    critical_threshold_pct = Column(Float, default=20.0)  # Critical at 20% deviation

    # Status
    is_active = Column(Boolean, default=True, nullable=False)

    # Notification settings
    notify_slack = Column(Boolean, default=True)
    notify_email = Column(Boolean, default=True)
    notify_whatsapp = Column(Boolean, default=False)
    notification_recipients = Column(JSONB, nullable=True)  # List of user IDs or emails

    # Timestamps
    created_at = Column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        nullable=False,
    )
    updated_at = Column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
        nullable=False,
    )
    created_by_user_id = Column(
        Integer, ForeignKey("users.id", ondelete="SET NULL"), nullable=True, index=True
    )

    # Relationships
    created_by = relationship("User", foreign_keys=[created_by_user_id])
    alerts = relationship(
        "PacingAlert", back_populates="target", cascade="all, delete-orphan"
    )

    __table_args__ = (
        Index("ix_targets_period", "period_start", "period_end"),
        Index("ix_targets_metric", "metric_type"),
        Index("ix_targets_active", "is_active"),
        Index("ix_targets_account", "account_id"),
        # was tenant-scoped; now global
        UniqueConstraint(
            "period_start",
            "period_end",
            "metric_type",
            "platform",
            "account_id",
            "campaign_id",
            name="uq_target_scope",
        ),
    )


# =============================================================================
# Daily KPI Model (Materialized)
# =============================================================================


class DailyKPI(Base):
    """
    Materialized daily KPIs for fast pacing calculations.
    Pre-aggregated from platform data and CRM metrics.
    """

    __tablename__ = "daily_kpis"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    date = Column(Date, nullable=False)

    # Scope (for granular tracking)
    platform = Column(String(50), nullable=True)  # null = all platforms
    campaign_id = Column(String(255), nullable=True)  # null = all campaigns
    account_id = Column(String(255), nullable=True)

    # Spend metrics
    spend_cents = Column(BigInteger, default=0, nullable=False)
    budget_cents = Column(BigInteger, nullable=True)  # Daily budget if set

    # Traffic metrics
    impressions = Column(BigInteger, default=0, nullable=False)
    clicks = Column(BigInteger, default=0, nullable=False)
    ctr = Column(Float, nullable=True)

    # Cost metrics
    cpm_cents = Column(BigInteger, nullable=True)
    cpc_cents = Column(BigInteger, nullable=True)
    cpa_cents = Column(BigInteger, nullable=True)
    cpl_cents = Column(BigInteger, nullable=True)

    # Conversion metrics (platform-reported)
    conversions = Column(Integer, default=0, nullable=False)
    leads = Column(Integer, default=0, nullable=False)
    purchases = Column(Integer, default=0, nullable=False)

    # Revenue metrics (platform-reported)
    revenue_cents = Column(BigInteger, default=0, nullable=False)
    roas = Column(Float, nullable=True)

    # CRM metrics (from HubSpot/Salesforce)
    crm_leads = Column(Integer, default=0, nullable=False)
    crm_mqls = Column(Integer, default=0, nullable=False)
    crm_sqls = Column(Integer, default=0, nullable=False)
    crm_opportunities = Column(Integer, default=0, nullable=False)
    crm_deals_won = Column(Integer, default=0, nullable=False)
    crm_pipeline_cents = Column(BigInteger, default=0, nullable=False)
    crm_won_revenue_cents = Column(BigInteger, default=0, nullable=False)

    # Computed CRM ROAS
    pipeline_roas = Column(Float, nullable=True)
    won_roas = Column(Float, nullable=True)

    # Quality metrics
    frequency = Column(Float, nullable=True)
    emq_score = Column(Float, nullable=True)

    # Day of week (for seasonality)
    day_of_week = Column(Integer, nullable=False)  # 0=Monday, 6=Sunday
    is_weekend = Column(Boolean, nullable=False)

    # Timestamps
    created_at = Column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        nullable=False,
    )
    updated_at = Column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
        nullable=False,
    )

    # Relationships

    __table_args__ = (
        Index("ix_daily_kpis_date", "date"),
        Index("ix_daily_kpis_platform_date", "platform", "date"),
        Index("ix_daily_kpis_campaign_date", "campaign_id", "date"),
        Index("ix_daily_kpis_account_date", "account_id", "date"),
        Index(
            "ix_daily_kpis_platform_account_date",
            "platform",
            "account_id",
            "date",
        ),
        # was tenant-scoped; now global
        UniqueConstraint(
            "date",
            "platform",
            "account_id",
            "campaign_id",
            name="uq_daily_kpi_scope",
        ),
    )


# =============================================================================
# Pacing Alert Model
# =============================================================================


class PacingAlert(Base):
    """
    Alert records for pacing issues.
    Tracks when targets are at risk of being missed.
    """

    __tablename__ = "pacing_alerts"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    target_id = Column(
        UUID(as_uuid=True),
        ForeignKey("targets.id", ondelete="CASCADE"),
        nullable=True,
        index=True,
    )

    # Alert details
    alert_type = Column(StrEnumType(AlertType), nullable=False)
    severity = Column(
        StrEnumType(AlertSeverity), nullable=False, default=AlertSeverity.WARNING
    )
    status = Column(
        StrEnumType(AlertStatus), nullable=False, default=AlertStatus.ACTIVE
    )

    # Alert message
    title = Column(String(255), nullable=False)
    message = Column(Text, nullable=False)
    details = Column(JSONB, nullable=True)  # Additional context

    # Metrics at time of alert
    current_value = Column(Float, nullable=True)
    target_value = Column(Float, nullable=True)
    projected_value = Column(Float, nullable=True)
    deviation_pct = Column(Float, nullable=True)

    # Pacing context
    pacing_date = Column(Date, nullable=False)
    days_remaining = Column(Integer, nullable=True)
    mtd_actual = Column(Float, nullable=True)
    mtd_expected = Column(Float, nullable=True)
    projected_eom = Column(Float, nullable=True)

    # Scope
    platform = Column(String(50), nullable=True)
    account_id = Column(String(255), nullable=True)  # Platform account ID
    campaign_id = Column(String(255), nullable=True)

    # Resolution
    resolved_at = Column(DateTime(timezone=True), nullable=True)
    resolved_by_user_id = Column(
        Integer, ForeignKey("users.id", ondelete="SET NULL"), nullable=True, index=True
    )
    resolution_notes = Column(Text, nullable=True)

    # Notification tracking
    notifications_sent = Column(JSONB, nullable=True)  # Track what was sent where

    # Timestamps
    created_at = Column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        nullable=False,
    )
    updated_at = Column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
        nullable=False,
    )

    # Relationships
    target = relationship("Target", back_populates="alerts")
    resolved_by = relationship("User", foreign_keys=[resolved_by_user_id])

    __table_args__ = (
        Index("ix_pacing_alerts_status", "status"),
        Index("ix_pacing_alerts_date", "pacing_date"),
        Index("ix_pacing_alerts_type", "alert_type"),
        Index("ix_pacing_alerts_target", "target_id"),
        Index("ix_pacing_alerts_account", "account_id"),
    )


# =============================================================================
# Forecast Model
# =============================================================================


class Forecast(Base):
    """
    Stored forecasts for trend analysis and auditing.
    Tracks daily, weekly, and EOM projections.
    """

    __tablename__ = "forecasts"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4)

    # Forecast metadata
    forecast_date = Column(Date, nullable=False)  # Date forecast was made
    forecast_for_date = Column(Date, nullable=False)  # Date being forecasted
    forecast_type = Column(String(50), nullable=False)  # daily, eom, custom

    # Scope
    platform = Column(String(50), nullable=True)
    account_id = Column(String(255), nullable=True)  # Platform account ID
    campaign_id = Column(String(255), nullable=True)

    # Metric being forecasted
    metric_type = Column(StrEnumType(TargetMetric), nullable=False)

    # Forecast values
    forecasted_value = Column(Float, nullable=False)
    confidence_lower = Column(Float, nullable=True)  # Lower bound (e.g., 90% CI)
    confidence_upper = Column(Float, nullable=True)  # Upper bound
    confidence_level = Column(
        Float, default=0.9
    )  # Confidence level (e.g., 0.9 for 90%)

    # Model info
    model_type = Column(String(50), nullable=False)  # ewma, linear, prophet, etc.
    model_params = Column(JSONB, nullable=True)  # Model parameters used

    # Accuracy tracking (filled in after actual is known)
    actual_value = Column(Float, nullable=True)
    error = Column(Float, nullable=True)  # actual - forecasted
    error_pct = Column(Float, nullable=True)  # Percentage error
    mape = Column(Float, nullable=True)  # Mean Absolute Percentage Error

    # Timestamps
    created_at = Column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        nullable=False,
    )

    # Relationships

    __table_args__ = (
        Index("ix_forecasts_date", "forecast_date"),
        Index("ix_forecasts_for_date", "forecast_for_date"),
        Index("ix_forecasts_metric", "metric_type"),
        Index("ix_forecasts_account", "account_id"),
    )


# =============================================================================
# Pacing Summary View (Helper)
# =============================================================================


class PacingSummary(Base):
    """
    Pacing summary snapshots (could be a materialized view).
    Stores current pacing status for quick dashboard loading.
    """

    __tablename__ = "pacing_summaries"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    target_id = Column(
        UUID(as_uuid=True),
        ForeignKey("targets.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )

    # Snapshot date
    snapshot_date = Column(Date, nullable=False)

    # Period info
    period_start = Column(Date, nullable=False)
    period_end = Column(Date, nullable=False)
    days_elapsed = Column(Integer, nullable=False)
    days_remaining = Column(Integer, nullable=False)
    days_total = Column(Integer, nullable=False)

    # Progress
    target_value = Column(Float, nullable=False)
    mtd_actual = Column(Float, nullable=False)
    mtd_expected = Column(Float, nullable=False)  # Pro-rated target

    # Pacing metrics
    pacing_pct = Column(Float, nullable=False)  # mtd_actual / mtd_expected * 100
    completion_pct = Column(Float, nullable=False)  # mtd_actual / target_value * 100

    # Projections
    projected_eom = Column(Float, nullable=True)
    projected_eom_lower = Column(Float, nullable=True)
    projected_eom_upper = Column(Float, nullable=True)

    # Gap analysis
    gap_to_target = Column(Float, nullable=True)  # target - projected_eom
    gap_pct = Column(Float, nullable=True)

    # Daily needed to hit target
    daily_needed = Column(
        Float, nullable=True
    )  # (target - mtd_actual) / days_remaining
    daily_average = Column(Float, nullable=True)  # mtd_actual / days_elapsed

    # Status flags
    on_track = Column(Boolean, nullable=False)
    at_risk = Column(Boolean, nullable=False)
    will_miss = Column(Boolean, nullable=False)

    # Timestamps
    created_at = Column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        nullable=False,
    )
    updated_at = Column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
        nullable=False,
    )

    # Relationships
    target = relationship("Target", foreign_keys=[target_id])

    __table_args__ = (
        Index("ix_pacing_summaries_date", "snapshot_date"),
        Index("ix_pacing_summaries_target", "target_id", "snapshot_date"),
        UniqueConstraint(
            "target_id", "snapshot_date", name="uq_pacing_summary_target_date"
        ),
    )

File: frontend/src/api/pacing.ts
/**
 * ADs Growth System - Pacing & Forecasting API
 *
 * Handles targets, pacing calculations, forecasts, and alerts.
 */

import { useQuery, useMutation, useQueryClient } from '@tanstack/react-query'
import { apiClient, ApiResponse } from './client'

// =============================================================================
// Types
// =============================================================================

export type TargetPeriod = 'daily' | 'weekly' | 'monthly' | 'quarterly' | 'yearly' | 'custom'
export type TargetMetric = 'spend' | 'revenue' | 'roas' | 'conversions' | 'cpa' | 'ctr' | 'impressions' | 'clicks'
export type AlertSeverity = 'info' | 'warning' | 'critical'
export type AlertType = 'underpacing' | 'overpacing' | 'target_at_risk' | 'target_missed' | 'target_exceeded'
export type AlertStatus = 'active' | 'acknowledged' | 'resolved' | 'dismissed'

export interface Target {
  id: string
  tenantId: number
  name: string
  description?: string
  metricType: TargetMetric
  targetValue: number
  periodType: TargetPeriod
  periodStart: string
  periodEnd: string
  platform?: string
  campaignId?: string
  minValue?: number
  maxValue?: number
  warningThresholdPct: number
  criticalThresholdPct: number
  isActive: boolean
  notifySlack: boolean
  notifyEmail: boolean
  notifyWhatsapp: boolean
  notificationRecipients: string[]
  createdAt: string
  updatedAt: string
}

export interface DailyKPI {
  id: string
  tenantId: number
  date: string
  platform?: string
  campaignId?: string
  spend: number
  revenue: number
  conversions: number
  impressions: number
  clicks: number
  roas: number
  cpa: number
  ctr: number
}

export interface PacingStatus {
  targetId: string
  targetName: string
  metricType: TargetMetric
  targetValue: number
  currentValue: number
  pacingPct: number
  projectedValue: number
  projectedPct: number
  daysRemaining: number
  daysElapsed: number
  dailyTarget: number
  dailyActual: number
  status: 'on_track' | 'ahead' | 'behind' | 'at_risk' | 'missed'
  trend: 'improving' | 'declining' | 'stable'
}

export interface PacingAlert {
  id: string
  tenantId: number
  targetId: string
  target?: Target
  alertType: AlertType
  severity: AlertSeverity
  status: AlertStatus
  message: string
  currentValue: number
  expectedValue: number
  deviationPct: number
  triggeredAt: string
  acknowledgedAt?: string
  resolvedAt?: string
  acknowledgedBy?: string
  resolvedBy?: string
  notes?: string
}

export interface Forecast {
  id: string
  tenantId: number
  targetId?: string
  metricType: TargetMetric
  platform?: string
  campaignId?: string
  forecastDate: string
  generatedAt: string
  horizon: number
  predictions: ForecastPrediction[]
  modelType: string
  accuracy?: number
  confidenceLevel: number
}

export interface ForecastPrediction {
  date: string
  predicted: number
  lowerBound: number
  upperBound: number
  confidencePct: number
}

export interface PacingSummary {
  id: string
  tenantId: number
  date: string
  period: TargetPeriod
  totalTargets: number
  onTrack: number
  ahead: number
  behind: number
  atRisk: number
  missed: number
  overallHealthScore: number
  topPerformers: string[]
  needsAttention: string[]
}

export interface CreateTargetRequest {
  name: string
  description?: string
  metricType: TargetMetric
  targetValue: number
  periodType: TargetPeriod
  periodStart: string
  periodEnd: string
  platform?: string
  campaignId?: string
  minValue?: number
  maxValue?: number
  warningThresholdPct?: number
  criticalThresholdPct?: number
  notifySlack?: boolean
  notifyEmail?: boolean
  notifyWhatsapp?: boolean
  notificationRecipients?: string[]
}

// =============================================================================
// API Functions
// =============================================================================

export const pacingApi = {
  // Targets
  getTargets: async (params?: { isActive?: boolean; metricType?: TargetMetric; periodType?: TargetPeriod }) => {
    const response = await apiClient.get<ApiResponse<Target[]>>('/pacing/targets', { params })
    return response.data.data
  },

  getTarget: async (targetId: string) => {
    const response = await apiClient.get<ApiResponse<Target>>(`/pacing/targets/${targetId}`)
    return response.data.data
  },

  createTarget: async (data: CreateTargetRequest) => {
    const response = await apiClient.post<ApiResponse<Target>>('/pacing/targets', data)
    return response.data.data
  },

  updateTarget: async (targetId: string, data: Partial<CreateTargetRequest>) => {
    const response = await apiClient.patch<ApiResponse<Target>>(`/pacing/targets/${targetId}`, data)
    return response.data.data
  },

  deleteTarget: async (targetId: string) => {
    const response = await apiClient.delete<ApiResponse<void>>(`/pacing/targets/${targetId}`)
    return response.data
  },

  // Pacing
  getPacingStatus: async (targetId: string) => {
    const response = await apiClient.get<ApiResponse<PacingStatus>>(`/pacing/targets/${targetId}/pacing`)
    return response.data.data
  },

  getAllPacingStatus: async () => {
    const response = await apiClient.get<ApiResponse<PacingStatus[]>>('/pacing/status')
    return response.data.data
  },

  getPacingSummary: async (period?: TargetPeriod) => {
    const response = await apiClient.get<ApiResponse<PacingSummary>>('/pacing/summary', { params: { period } })
    return response.data.data
  },

  // Daily KPIs
  getDailyKPIs: async (params: { startDate: string; endDate: string; platform?: string; campaignId?: string }) => {
    const response = await apiClient.get<ApiResponse<DailyKPI[]>>('/pacing/kpis/daily', { params })
    return response.data.data
  },

  // Forecasts
  getForecast: async (targetId: string, horizon?: number) => {
    const response = await apiClient.get<ApiResponse<Forecast>>(`/pacing/targets/${targetId}/forecast`, {
      params: { horizon },
    })
    return response.data.data
  },

  getMetricForecast: async (params: { metricType: TargetMetric; platform?: string; horizon?: number }) => {
    const response = await apiClient.get<ApiResponse<Forecast>>('/pacing/forecast', { params })
    return response.data.data
  },

  generateForecast: async (targetId: string, horizon?: number) => {
    const response = await apiClient.post<ApiResponse<Forecast>>(`/pacing/targets/${targetId}/forecast/generate`, {
      horizon,
    })
    return response.data.data
  },

  // Alerts
  getAlerts: async (params?: { status?: AlertStatus; severity?: AlertSeverity; targetId?: string }) => {
    const response = await apiClient.get<ApiResponse<PacingAlert[]>>('/pacing/alerts', { params })
    return response.data.data
  },

  getAlert: async (alertId: string) => {
    const response = await apiClient.get<ApiResponse<PacingAlert>>(`/pacing/alerts/${alertId}`)
    return response.data.data
  },

  acknowledgeAlert: async (alertId: string, notes?: string) => {
    const response = await apiClient.post<ApiResponse<PacingAlert>>(`/pacing/alerts/${alertId}/acknowledge`, { notes })
    return response.data.data
  },

  resolveAlert: async (alertId: string, notes?: string) => {
    const response = await apiClient.post<ApiResponse<PacingAlert>>(`/pacing/alerts/${alertId}/resolve`, { notes })
    return response.data.data
  },

  dismissAlert: async (alertId: string, notes?: string) => {
    const response = await apiClient.post<ApiResponse<PacingAlert>>(`/pacing/alerts/${alertId}/dismiss`, { notes })
    return response.data.data
  },

  getAlertStats: async () => {
    const response = await apiClient.get<ApiResponse<{
      total: number
      active: number
      acknowledged: number
      bySeverity: Record<AlertSeverity, number>
      byType: Record<AlertType, number>
    }>>('/pacing/alerts/stats')
    return response.data.data
  },
}

// =============================================================================
// React Query Hooks
// =============================================================================

// Targets
export function useTargets(params?: { isActive?: boolean; metricType?: TargetMetric; periodType?: TargetPeriod }) {
  return useQuery({
    queryKey: ['pacing', 'targets', params],
    queryFn: () => pacingApi.getTargets(params),
  })
}

export function useTarget(targetId: string) {
  return useQuery({
    queryKey: ['pacing', 'targets', targetId],
    queryFn: () => pacingApi.getTarget(targetId),
    enabled: !!targetId,
  })
}

export function useCreateTarget() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: pacingApi.createTarget,
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['pacing', 'targets'] })
    },
  })
}

export function useUpdateTarget() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: ({ targetId, data }: { targetId: string; data: Partial<CreateTargetRequest> }) =>
      pacingApi.updateTarget(targetId, data),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['pacing', 'targets'] })
    },
  })
}

export function useDeleteTarget() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: pacingApi.deleteTarget,
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['pacing', 'targets'] })
    },
  })
}

// Pacing
export function usePacingStatus(targetId: string) {
  return useQuery({
    queryKey: ['pacing', 'status', targetId],
    queryFn: () => pacingApi.getPacingStatus(targetId),
    enabled: !!targetId,
    refetchInterval: 60000, // Refresh every minute
  })
}

export function useAllPacingStatus() {
  return useQuery({
    queryKey: ['pacing', 'status', 'all'],
    queryFn: pacingApi.getAllPacingStatus,
    refetchInterval: 60000,
  })
}

export function usePacingSummary(period?: TargetPeriod) {
  return useQuery({
    queryKey: ['pacing', 'summary', period],
    queryFn: () => pacingApi.getPacingSummary(period),
    refetchInterval: 60000,
  })
}

// Daily KPIs
export function useDailyKPIs(params: { startDate: string; endDate: string; platform?: string; campaignId?: string }) {
  return useQuery({
    queryKey: ['pacing', 'kpis', 'daily', params],
    queryFn: () => pacingApi.getDailyKPIs(params),
    enabled: !!params.startDate && !!params.endDate,
  })
}

// Forecasts
export function useForecast(targetId: string, horizon?: number) {
  return useQuery({
    queryKey: ['pacing', 'forecast', targetId, horizon],
    queryFn: () => pacingApi.getForecast(targetId, horizon),
    enabled: !!targetId,
  })
}

export function useMetricForecast(params: { metricType: TargetMetric; platform?: string; horizon?: number }) {
  return useQuery({
    queryKey: ['pacing', 'forecast', 'metric', params],
    queryFn: () => pacingApi.getMetricForecast(params),
  })
}

export function useGenerateForecast() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: ({ targetId, horizon }: { targetId: string; horizon?: number }) =>
      pacingApi.generateForecast(targetId, horizon),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['pacing', 'forecast'] })
    },
  })
}

// Alerts
export function usePacingAlerts(params?: { status?: AlertStatus; severity?: AlertSeverity; targetId?: string }) {
  return useQuery({
    queryKey: ['pacing', 'alerts', params],
    queryFn: () => pacingApi.getAlerts(params),
    refetchInterval: 30000, // Refresh every 30 seconds
  })
}

export function usePacingAlert(alertId: string) {
  return useQuery({
    queryKey: ['pacing', 'alerts', alertId],
    queryFn: () => pacingApi.getAlert(alertId),
    enabled: !!alertId,
  })
}

export function useAcknowledgeAlert() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: ({ alertId, notes }: { alertId: string; notes?: string }) =>
      pacingApi.acknowledgeAlert(alertId, notes),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['pacing', 'alerts'] })
    },
  })
}

export function useResolveAlert() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: ({ alertId, notes }: { alertId: string; notes?: string }) =>
      pacingApi.resolveAlert(alertId, notes),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['pacing', 'alerts'] })
    },
  })
}

export function useDismissAlert() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: ({ alertId, notes }: { alertId: string; notes?: string }) =>
      pacingApi.dismissAlert(alertId, notes),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['pacing', 'alerts'] })
    },
  })
}

export function useAlertStats() {
  return useQuery({
    queryKey: ['pacing', 'alerts', 'stats'],
    queryFn: pacingApi.getAlertStats,
    refetchInterval: 30000,
  })
}
